lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -27,6 +130,274 @@
+
+        return final_routing_table
+
+    def _sync_routing_tables(self) -> None:
+        """Syncs the internal routing table with all of the requirer's app databags"""
+        ip_router_relations = self.model.relations["ip-router"]
+        for relation in ip_router_relations:
+            relation.data[self.charm.app].update(
+                {"networks": json.dumps(self.get_flattened_routing_table())}
+            )
+
+
+class RouterRequires(Object):
+    """ip-router requirer class to be instantiated by charms that require routing
+
+    This class provides methods to request a new network, and read the available
+    network from the router providers. These should be used exclusively to
+    interact with the relation.
+
+    Attributes:
+        charm: The Charm object that instantiates this class.
+    """
+
+    def __init__(self, charm: CharmBase, relationship_name: str = "ip-router"):
+        super().__init__(charm, relationship_name)
+        self.charm = charm
+        self.relationship_name = relationship_name
+
+    def request_network(
+        self,
+        networks: List[Network],
+    ) -> None:
+        """Requests a new network interface from the ip-router provider
+
+        The interfaces must be valid according to `_network_is_valid`. Multiple
+        calls to this function will replace the previously requested networks,
+        so all of the networks required must be given with each call.
+
+        Arguments:
+            networks:
+                A list containing the desired networks of the type `Network`.
+
+        Raises:
+            RuntimeError:
+                No ip-router relation exists yet or validation of one
+            or more of the networks failed.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        ip_router_relations = self.model.relations.get(self.relationship_name)
+        if len(ip_router_relations) == 0:
+            raise RuntimeError("No ip-router relation exists yet.")
+
+        for network_request in networks:
+            self._validate_network(network_request, networks)
+
+        # Place it in the databags
+        for relation in ip_router_relations:
+            relation.data[self.charm.app].update({"networks": json.dumps(networks)})
+
+    def get_all_networks(self) -> List[Network]:
+        """Fetches combined routing tables made available by ip-router providers
+
+        Args:
+            None
+        Returns:
+            A list of objects of type `Network`. This list contains networks
+            from all ip-router providers that are integrated with the charm.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        router_relations = self.model.relations.get("ip-router")
+        all_networks = []
+        for relation in router_relations:
+            if networks := relation.data[relation.app].get("networks"):
+                all_networks.extend(json.loads(networks))
```

Comment:
I think we should catch the exceptions here, log an error and ignore the broken networks.

Reply:
Done